### PR TITLE
feat(core,components): Edit button improvements

### LINKED ISSUE #624: Consolidate edit form APIs
## User Story
As a developer
I want one consistent API for creating edit forms
So that I can work more efficiently

## Background

We currently have 2 converging API's for creating forms:
- withEditButton
- withCompundForm + withEditFormSnippet

These should be consolidated:

- withEditButton should always use withCompundForm. This would allow sub-compnents to add fields to the edit forms of any component.  Currently, withEditButton [can *optionally* use a compound form](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-core/src/withEditButton.tsx#L116-L124), although this is not complete:
    - when toggled on it's really just a wrapper around `withCompundForm`, whereas it should also add a snippet if `renderForm` is supplied.)
    - Also, in the current implementation, [you cannot specify options a a function when using a compound form](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-core/src/withEditButton.tsx#L116-L118), which is bad DX.
- withEditButton should allow you to control attributes of the [menu options definition](https://github.com/johnsonandjohnson/Bodiless-JS/blob/master/packages/bodiless-core/src/Types/PageContextProviderTypes.tsx#L22-L24), not just the menu option.  For example, it's not current'y possible to create a peer context using withEditButton.  This is possible for withCompoundForm.
- withCompoundForm currently doesn't provide any grouping mechanism for its button (by contrast, [withEditButton will  soon automatically add a group and provide control over the name and merge behavior](https://github.com/johnsonandjohnson/Bodiless-JS/pull/597/files#diff-c9d505791fc5ab51954a09deaa490ca2898166d1d4b9174491b19169e3df4f47R105-R116).

A few notes:
- we can't do away with "withCompundForm" completely, bc it can be used for things other than edit forms.
- it's unclear what the best API signature is.  withCompundForm currently takes a MenuOptionsDefinition, which allows it to modify properties of the context.  This is definitely more flexible - and also has the advantage of being the same as "withMenuOptions".  On the other hand, it's more verbose, and extending this to `withEditButton` would mean refactoring all the consumers, and raise the question of whether we should also extend it to "asBodilessComponent"...

### Options and Overrides

Another consideration is how this will affect asBodilessComponent and the 'useOverrides' api.   Perhaps 'useOverrides' should be replaced by an 'overrides' object which has a 'useMenuOptionOverrides' key?  This would allow overriding of context-level properties (like 'name', 'type' or 'peer'.  OTOH - would it be better, instead, to let components do this themselves?

Another proposa is to supply overrides (and maybe even options) via a prop.  So instead of, eg 
```
asBodilessLink(nodeKey, defaultData, useOverrides)
```
we'd have
```
flow(
  asBodilessLink(nodeKey, defaultData)
  addProps({ useOverrides }),
)
```

A problem here is stacked enhancements, eg:
```
flow(
  asBodilessFoo(),
  asBodilessBar(),
  addProps({ useOverrides }),
)
```

Should the overrides be used for Foo or Bar.  One Proposed solution is to have asBodilessComponent add a design key, eg:
```
const asFooAndBar = flow(
  asBodilessFoo(),
  asBodilessBar(),
  withDesign({
    Foo: addProps({ useFooOverrides }),
    Bar: addProps({ useBarOverrides }),
  }),
);
```

Note that in this solution, if the wrapped component were not itself designable, we'd have to terminate the design, eg
```
flow(
  withoutProps('design'),
  asFooAndBar,
)('span');
```

Another soluiton would be to require recomposition - eg
```
flow(
  asBodilessFoo(),
  addProps({ useFooOverrides }),
  asBodilessBar(),
  addProps({ useBarOverrides }),
);
```

Note that in either solution, addProps itself is probalby not sufficient, as we will want to merge overrides, so we need something like
```
const withOverrides = (extOverrides) => Component => ({ useOverrides, ...rest }) => {
  const useOverrrides$ = props => ({ ...useOverrides(props), ...extOverrides(props) });
  return <Compoent useOverrides={useOverrides$} {...rest} />
}
```

In this scenario, we'd probably refactor all the core hocs which currently use options to get theri options from a prop, eg

```
const withMenuOptions = (options) => Component: => props => {
  const { useOverides, ...rest } = props;
  const overrides = useOverrides(props);
  const finalOptions = {...options, ...overrides}
  ...
}
```

Or even

```
const withMenuOptions$ = (Component) => props => {
  const { useOptions, ...rest } = props;
  const options = useOptions(props);
  ,,,
}

const withMenuOptions = (options) => flow(
  withMenuOptions$,
  withOptions(options),
);
```

### LINKED ISSUE #794: Allow ability to add buttons to be added to toolbar
## User Story
As a site builder
I want my component to add buttons to the admin toolbar
via an api interface so that they visible in the toolbar and not just in context.

## Acceptance Criteria
AC1. It is possible to configure a component to add buttons to the (global admin) toolbar.
AC2: In the tool bar, when selected the Page button opens a sub panel for "Page" with the "New Page" icon and label
AC3. When the user selects the "New Page" button the "Add Blank Page" form appears with same behavior as it does now.
AC4. There is a test case on the Chameleon page on the Test site, in which a chameleon adds a button to the tool bar to toggle its visibility 

## Notes
- In scope of this, we need a way for any component, however deeply nested, to add a global button which is always visible - and probablly we should simplify so that
  - All "global" buttons are always visible (peers of the root context)
  - No button is both global and local

### PR DESCRIPTION
**Pull request by wodenx**
## Changes
- Use compound form for all edit buttons (#624)
- Allow chameleons (and other components) to place buttons on main toolbar (#794)
- Allow overriding MenuOptionsDefinition via useOverrides
- Add "Page" submenu and put New, SEO and Share buttons there.

## Test Instructions
- Test the layout functionality on the chameleon page
- Regression testing of edit forms (link, image, SEO, Share).

## Related Issues
Closes #624
Closes #794

BREAKING CHANGE: Structure of compound form data has changerd.

Fields in compound forms are now namespaced by the snippet id.

Since edit forms (those created by withEditButton or asBodilessComponent) are now
compound forms, this change affects them as well.

### CODE REVIEW

File: packages/bodiless-components/src/Chameleon/withChameleonButton.tsx [hunk 1]
```diff
@@ -5,20 +5,23 @@
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  * http://www.apache.org/licenses/LICENSE-2.0
- * Unless required by applicable law or agreed to in writing, software
+ * Unless required by applicable law or agreed to in writingpi, software
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/Chameleon/withChameleonButton.tsx:8]: fixed

File: packages/bodiless-components/src/Chameleon/withChameleonButton.tsx [hunk 2]
```diff
@@ -112,18 +115,29 @@ const withChameleonButton = <P extends object, D extends object>(
...
+    name: 'Chameleon',
+    ...pick(useOverrides(props), 'root', 'peer'),
+  });
+  const useHasLocalContext = (props: P & EditButtonProps<D>): boolean => {
+    const def = useMenuOptionsDefinition(props);
+    const isRoot = def.root || (def.peer && !useEditContext().parent);
+    return !isRoot;
+  };
+  return flow(
+    flowIf(useHasLocalContext)(
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/Chameleon/withChameleonButton.tsx:136]: per discussion, we don't want a context activator if this is part of the global context.

File: packages/bodiless-components/src/Image.tsx [hunk 3]
```diff
@@ -169,23 +169,25 @@ export function DropZonePlugin({ formApi, targetFieldName, ui }: {
     </MasterWrapper>
   );
 }
-DropZonePlugin.defaultProps = {
-  ui: {},
-};
+// DropZonePlugin.defaultProps = {
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-components/src/Image.tsx:172]: yep

File: packages/bodiless-core/src/ContextMenu/ContextSubMenu.tsx [hunk 4]
```diff
@@ -48,13 +57,23 @@ const ContextSubMenu: FC<IContextMenuItemProps> = props => {
...
         <SubMenu>
-          {children}
+          <SubMenuChildren />
         </SubMenu>
       </FormChrome>
     </ContextMenuBase>
   );
   const newOption = { ...option, handler };
-  return <ContextMenuItem option={newOption} name={option.name} {...rest} />;
+  return (
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/ContextMenu/ContextSubMenu.tsx:66]: i think i explained it sufficiently, no?

File: packages/bodiless-core/src/PageContextProvider.tsx [hunk 5]
```diff
@@ -138,14 +141,26 @@ export const withMenuOptions = <P extends object>(def$: MenuOptionsDefinition$<P
...
-        useRegisterMenuOptions({ getMenuOptions, ...rest });
+      if (root || peer) {
+        // Note we do not expect root or peer to change between renders, so it's
+        // ok to use hooks inside this condition.
+        let context = useEditContext();
+        if (root) {
+          while (context.parent) context = context.parent;
+        }
+        // Set the default menu placement based on whether we are attaching to the root context.
+        const getMenuOptions = options && useGetter(options.map(o => ({
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/PageContextProvider.tsx]: added some comments.

File: packages/bodiless-core/src/asBodilessComponent.tsx [hunk 6]
```diff
@@ -135,9 +142,11 @@ const asBodilessComponent = <P extends object, D extends object>(options: Option
       ),
       ifEditable(
         withEditButton(editButtonOptions),
-        withContextActivator(activateEvent),
-        withLocalContextMenu,
-        Wrapper ? withActivatorWrapper(activateEvent, Wrapper) : identity,
+        flowIf(useHasLocalContext)(
```
[the remark this answers is not in the document]
Comment (wodenx) [reply packages/bodiless-core/src/asBodilessComponent.tsx]: discussed.

**Review by e2thex — DISMISSED** [review]

**Review by e2thex — APPROVED** [review]

**Review by awm086 — APPROVED** [review]

**Pull request merged**